Program.cs

```diff
@@ -1,178 +1,264 @@
+
+async Task DownloadHandler(InvocationContext invocationContext)
+{
+    var booruValue = invocationContext.ParseResult.GetValueForOption(booru)!;
+    var tagsValue = invocationContext.ParseResult.GetValueForOption(tags)!;
+    var downloadPathValue = invocationContext.ParseResult.GetValueForOption(downloadPath)!;
+    var cunnyApiUrlValue = invocationContext.ParseResult.GetValueForOption(cunnyApiUrl)!;
+    var amountValue = invocationContext.ParseResult.GetValueForOption(amount)!;
+    var skipValue = invocationContext.ParseResult.GetValueForOption(skip)!;
+    var excludeTagsValue = invocationContext.ParseResult.GetValueForOption(excludeTags);
+    var proxyOptionValue = invocationContext.ParseResult.GetValueForOption(proxyOption);
+    var threadsValue = invocationContext.ParseResult.GetValueForOption(threads)!;
+
+    if (proxyOptionValue is not null)
+    {
+        var proxy = new WebProxy(proxyOptionValue);
+        var handler = new HttpClientHandler {Proxy = proxy};
+        var client = new HttpClient(handler);
+        Globals.Client = client;
     }
-    private readonly static object logsLock = new();
-    private readonly static object downloadThreadsLock = new();
-    private static readonly List<Task> downloadThreads = new();
-}
\ No newline at end of file
+
+    var results = await CunnyApiClient.Get(cunnyApiUrlValue, booruValue, tagsValue, amountValue, skipValue);
+    if (results is null)
+        return;
+
+    if (excludeTagsValue is not null)
+        results = new List<CunnyJsonElement>(results.Where(x => !x.Tags.Any(excludeTagsValue.Contains)));
+
+    Globals.Logs.CollectionChanged += (_, _) =>
+        Console.WriteLine(
+            $"\u001b[s\u001b[22;37m[\u001b[32m{progress}\u001b[37m/\u001b[0m{amountValue}\u001b[37m]\u001b[0m {Globals.Logs[^1]}\u001b[0J\u001b[u");
+
+    Parallel.ForEach(results, new ParallelOptions { MaxDegreeOfParallelism = threadsValue }, itemValue =>
+    {
+        var dirPath = Path.Combine(downloadPathValue, tagsValue, new Uri(itemValue.ImageURL).Host);
+        var filePath = Path.Combine(dirPath, $"{itemValue.ID}-{itemValue.Width}x{itemValue.Height}{Path.GetExtension(itemValue.ImageURL)}");
+        var filePathPart = $"{filePath}.part";
+
+        if (File.Exists(filePath))
+        {
+            Globals.Logs.Add($"Skipping \u001b[22;37m{Path.GetFileName(filePath)}\u001b[0m");
+            progress++;
+            return;
+        }
+
+        if(!Directory.Exists(dirPath))
+            Directory.CreateDirectory(dirPath);
+
+        Globals.Logs.Add($"Downloading \u001b[22;37m{Path.GetFileName(filePath)}\u001b[0m");
+        using var downloadStream = Globals.Client.GetStreamAsync(itemValue.ImageURL).Result;
+        using var writeStream = File.Open(filePathPart, FileMode.Create);
+        downloadStream.CopyTo(writeStream);
+        File.Move(filePathPart, filePath);
+        Globals.Logs.Add($"Saved \u001b[22;37m{Path.GetFileName(filePath)}\u001b[0m");
+        progress++;
+    });

```

Comment:
Can you add this QoL feature?

```suggestion
    });

    Console.WriteLine($"\u001b[u\u001b[0JSaved to \u001b[22;37m{downloadPathValue}\u001b[0m");
```